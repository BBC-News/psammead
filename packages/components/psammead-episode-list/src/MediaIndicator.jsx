import React from 'react';
import styled from '@emotion/styled';
import { string, number } from 'prop-types';
import { GEL_SPACING, GEL_SPACING_DBL } from '@bbc/gel-foundations/spacings';
import {
  GEL_GROUP_1_SCREEN_WIDTH_MAX,
  GEL_GROUP_2_SCREEN_WIDTH_MIN,
} from '@bbc/gel-foundations/breakpoints';

const Wrapper = styled.div`
  display: inline-block;
  width: ${props => props.size};
  height: ${props => props.size};
  @media (max-width: ${GEL_GROUP_1_SCREEN_WIDTH_MAX}) {
    margin-left: ${props => (props.dir === 'ltr' ? 0 : GEL_SPACING_DBL)};
    margin-right: ${props => (props.dir === 'ltr' ? GEL_SPACING_DBL : 0)};
    padding-top: ${props => (props.dir === 'ltr' ? 0 : '0.25rem')};
  }
  @media (min-width: ${GEL_GROUP_2_SCREEN_WIDTH_MIN}) {
    margin-left: ${props =>
      props.dir === 'ltr' ? GEL_SPACING : GEL_SPACING_DBL};
    margin-right: ${props =>
      props.dir === 'ltr' ? GEL_SPACING_DBL : GEL_SPACING};
    padding-top: ${props => (props.dir === 'ltr' ? 0 : '0.25rem')};
  }
  vertical-align: top;
`;

const MediaIndicator = ({ size, dir }) => (
  <Wrapper
    aria-hidden="true"
    size={size}
    dir={dir}
    dangerouslySetInnerHTML={{
      __html: `
      <svg class="rounded-play-button" focusable="false" width="40" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 40 40">
        <circle class="rounded-play-button__circle" fill="none" stroke="currentColor" stroke-width="1.765" stroke-miterlimit="10" cx="20" cy="20" r="19.119"/>
        <polygon class="rounded-play-button__triangle" fill="currentColor" points="15.799,26.939 27.062,20.008 15.799,13.078 "/>
      </svg>
      `,
    }}
  />
);

MediaIndicator.propTypes = {
  size: number.isRequired,
  dir: string.isRequired,
};

export default MediaIndicator;
