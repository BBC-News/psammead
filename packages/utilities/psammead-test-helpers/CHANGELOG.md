# Psammead Test Helpers Changelog

| Version | Description |
|---------|-------------|
| 0.3.0   | [PR#224](https://github.com/BBC-News/psammead/pull/224) Adds a helper method for testing package export values and uses it to test its own exports. |
| 0.2.1   | [PR#255](https://github.com/BBC-News/psammead/pull/255) Bumps package version to fix what appears to be an erroneously-published version. |
| 0.1.2   | [PR#212](https://github.com/BBC-News/psammead/pull/212) Update package description and README. |
| 0.1.1   | [PR#173](https://github.com/BBC-News/psammead/pull/173) Update PRs welcome link |
| 0.1.0   | [PR#80](https://github.com/BBC-News/psammead/pull/80) Create initial package, pulled in from [simorgh](https://github.com/BBC-News/simorgh). |
