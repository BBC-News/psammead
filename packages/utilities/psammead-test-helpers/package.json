{
  "name": "@bbc/psammead-test-helpers",
  "version": "0.3.0",
  "main": "dist/index.js",
  "description": "A collection of helper methods for implementing Jest snapshot tests for styled-components, required by many Psammead components.",
  "repository": {
    "type": "git",
    "url": "https://github.com/BBC-News/psammead/tree/latest/packages/utilities/psammead-test-helpers"
  },
  "author": {
    "name": "Psammead Maintainers",
    "email": "PsammeadMaintainers@bbc.co.uk"
  },
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/bbc/psammead/issues"
  },
  "homepage": "https://github.com/BBC-News/psammead/blob/latest/packages/utilities/psammead-test-helpers/README.md",
  "keywords": [
    "bbc",
    "utilities",
    "test",
    "helpers",
    "jest",
    "snapshot"
  ],
  "dependencies": {
    "jest-styled-components": "^6.3.1",
    "react-test-renderer": "^16.6.3"
  }
}
