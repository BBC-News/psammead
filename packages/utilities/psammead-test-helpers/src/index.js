import renderer from 'react-test-renderer';
import 'jest-styled-components';
import ShallowRenderer from 'react-test-renderer/shallow';

export const shouldMatchSnapshot = (title, component) => {
  it(title, () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
};

export const shallowRender = component => {
  const shallowRenderer = new ShallowRenderer();
  shallowRenderer.render(component);
  const result = shallowRenderer.getRenderOutput();

  return result;
};

export const shouldShallowMatchSnapshot = (title, component) => {
  it(title, () => {
    const tree = shallowRender(component);
    expect(tree).toMatchSnapshot();
  });
};

export const isNull = (title, component) => {
  it(title, () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toBeNull();
  });
};

const errorIfMissingKey = (keys, object, message) => {
  keys.forEach(key => {
    if (!(key in object)) {
      throw new Error(`Missing value '${key}' in ${message}.`);
    }
  });
};

const checkKeysExistInBothObjects = (object1, object2, message1, message2) => {
  const object1Keys = Object.keys(object1);
  const object2Keys = Object.keys(object2);

  errorIfMissingKey(object1Keys, object2, message2);
  errorIfMissingKey(object2Keys, object1, message1);
};

const checkTypesOfExports = (
  actualExportsByName,
  actualExports,
  expectedExports,
  utilityName,
) => {
  actualExportsByName.forEach(actualExportName => {
    const actualExportValue = actualExports[utilityName][actualExportName];
    const expectedExport = expectedExports[utilityName][actualExportName];
    const typeCheck = typeof actualExportValue === expectedExport; // eslint-disable-line valid-typeof

    // if this fails it is likely that an export is missing from the unit test expectation
    expect(typeCheck).toBe(true);
  });
};

export const testUtilityPackages = (
  actualExports,
  expectedExports,
  packageName,
) => {
  const actualUtilities = Object.keys(actualExports);

  // check if the actual exported file has defined expected values to match and that the expected values are actually exported
  checkKeysExistInBothObjects(
    actualExports,
    expectedExports,
    `the actual utilities for '${packageName}'`,
    `the expected utilities for '${packageName}'`,
  );

  actualUtilities.forEach(utilityName => {
    const actualExportsByName = Object.keys(actualExports[utilityName]);

    // check if each of the actual exported consts have an expected value to match and that the expected values are actually exported
    checkKeysExistInBothObjects(
      actualExports[utilityName],
      expectedExports[utilityName],
      `the actual export for '${packageName}/${utilityName}'`,
      `the expected export for '${packageName}/${utilityName}'`,
    );

    checkTypesOfExports(
      actualExportsByName,
      actualExports,
      expectedExports,
      utilityName,
    );
  });
};
