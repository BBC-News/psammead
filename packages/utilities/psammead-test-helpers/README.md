# psammead-test-helpers &middot; [![GitHub license](https://img.shields.io/badge/license-Apache%202.0-blue.svg)](https://github.com/BBC-News/psammead/blob/latest/LICENSE) [![npm version](https://img.shields.io/npm/v/@bbc/psammead-test-helpers.svg)](https://www.npmjs.com/package/@bbc/psammead-test-helpers) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/BBC-News/psammead/blob/latest/CONTRIBUTING.md)

This package is a collection helper methods used for snapshot testing, with support for styled components.

## Installation

```jsx
npm install react react-dom @bbc/psammead-test-helpers --save-dev
```

## Usage

```jsx
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';

shouldMatchSnapshot(
  'should render correctly',
  <h1>Hello World</h1>,
);
```

## Functions

| Name                       | Arguments        | Description  |
|:---------------------------|:-----------------|:-------------|
| shouldMatchSnapshot        | title, component | Renders the component using react-test-renderer, converts it to JSON and asserts that it matches the given snapshot, which will be saved in the `__snapshots__` directory. The first argument `title` is the title for the test. |
| shallowRender              | component        |  Shallow renders the component using react-test-renderer and returns the render output |
| shouldShallowMatchSnapshot | title, component | Shallow renders the component using react-test-renderer and asserts that it matches the given snapshot, which will be saved in the `__snapshots__` directory. The first argument `title` is the title for the test. |
| isNull                     | title, component    | Renders the component using react-test-renderer, converts it to JSON and asserts that it is null. The first argument `title` is the title for the test. |

## Roadmap

There is currently a plan to migrate from `react-test-renderer` to `enzyme` as it provides a nicer higher-level API over the same functionality, as well as having support for more granular unit tests.

## Contributing

Psammead is completely open source. We are grateful for any contributions, whether they be new components, bug fixes or general improvements. Please see our primary contributing guide which can be found at [the root of the Psammead respository](https://github.com/BBC-News/psammead/blob/latest/CONTRIBUTING.md).

### [Code of Conduct](https://github.com/BBC-News/psammead/blob/latest/CODE_OF_CONDUCT.md)

We welcome feedback and help on this work. By participating in this project, you agree to abide by the [code of conduct](https://github.com/BBC-News/psammead/blob/latest/CODE_OF_CONDUCT.md). Please take a moment to read it.

### License

Psammead is [Apache 2.0 licensed](https://github.com/BBC-News/psammead/blob/latest/LICENSE).
