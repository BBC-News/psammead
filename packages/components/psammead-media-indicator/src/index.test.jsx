import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import MediaIndicator from './index';

describe('MediaIndicator', () => {
  shouldMatchSnapshot(
    'should render video by default',
    <MediaIndicator offscreenText="Video" service="news" />,
  );

  shouldMatchSnapshot(
    'should render video indicator correctly',
    <MediaIndicator
      duration="2:15"
      datetime="PT2M15S"
      offscreenText="Video"
      type="video"
      service="news"
    />,
  );

  shouldMatchSnapshot(
    'should render video correctly without duration details',
    <MediaIndicator offscreenText="Video" type="video" service="news" />,
  );

  shouldMatchSnapshot(
    'should render audio indicator correctly',
    <MediaIndicator
      duration="2:15"
      datetime="PT2M15S"
      offscreenText="Audio"
      type="audio"
      service="news"
    />,
  );

  shouldMatchSnapshot(
    'should render audio correctly without duration details',
    <MediaIndicator offscreenText="Audio" type="audio" service="news" />,
  );

  shouldMatchSnapshot(
    'should render photogallery correctly',
    <MediaIndicator type="photogallery" service="news" />,
  );
});

describe('MediaIndicator - Top Story', () => {
  shouldMatchSnapshot(
    'should render a top story audio promo render correctly',
    <MediaIndicator
      duration="2:15"
      datetime="PT2M15S"
      offscreenText="Audio"
      type="audio"
      topStory
      service="news"
    />,
  );
});
