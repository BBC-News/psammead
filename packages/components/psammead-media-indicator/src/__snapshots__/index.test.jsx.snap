// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaIndicator should render audio correctly without duration details 1`] = `
.c3 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
}

.c2 {
  vertical-align: middle;
  margin: 0 0.25rem;
  fill: #222222;
  width: 0.8125rem;
  height: 0.75rem;
}

.c0 {
  padding: 0.5rem 0.25rem;
  background-color: #FFFFFF;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  color: #222222;
  height: 2rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

<div
  aria-hidden="true"
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      class="c2"
      focusable="false"
      height="12px"
      viewBox="0 0 13 12"
      width="13px"
    >
      <path
        d="M9.021 1.811l-.525.525c.938.938 1.5 2.25 1.5 3.675s-.563 2.738-1.5 3.675l.525.525c1.05-1.087 1.725-2.55 1.725-4.2s-.675-3.112-1.725-4.2z"
      />
      <path
        d="M10.596.199l-.525.562c1.35 1.35 2.175 3.225 2.175 5.25s-.825 3.9-2.175 5.25l.525.525c1.5-1.462 2.4-3.525 2.4-5.775s-.9-4.312-2.4-5.812zM6.996 1.511l-2.25 2.25H.996v4.5h3.75l2.25 2.25z"
      />
    </svg>
    <span
      class="c3"
    >
      Audio
    </span>
  </div>
</div>
`;

exports[`MediaIndicator should render audio indicator correctly 1`] = `
.c3 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
}

.c2 {
  vertical-align: middle;
  margin: 0 0.25rem;
  fill: #222222;
  width: 0.8125rem;
  height: 0.75rem;
}

.c0 {
  padding: 0.5rem 0.25rem;
  background-color: #FFFFFF;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  color: #222222;
  height: 2rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

.c4 {
  vertical-align: middle;
  margin: 0 0.25rem;
}

<div
  aria-hidden="true"
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      class="c2"
      focusable="false"
      height="12px"
      viewBox="0 0 13 12"
      width="13px"
    >
      <path
        d="M9.021 1.811l-.525.525c.938.938 1.5 2.25 1.5 3.675s-.563 2.738-1.5 3.675l.525.525c1.05-1.087 1.725-2.55 1.725-4.2s-.675-3.112-1.725-4.2z"
      />
      <path
        d="M10.596.199l-.525.562c1.35 1.35 2.175 3.225 2.175 5.25s-.825 3.9-2.175 5.25l.525.525c1.5-1.462 2.4-3.525 2.4-5.775s-.9-4.312-2.4-5.812zM6.996 1.511l-2.25 2.25H.996v4.5h3.75l2.25 2.25z"
      />
    </svg>
    <span
      class="c3"
    >
      Audio
    </span>
    <time
      class="c4"
      datetime="PT2M15S"
    >
      2:15
    </time>
  </div>
</div>
`;

exports[`MediaIndicator should render video by default 1`] = `
.c3 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
}

.c2 {
  vertical-align: middle;
  margin: 0 0.25rem;
  fill: #222222;
  width: 0.75rem;
  height: 0.75rem;
}

.c0 {
  padding: 0.5rem 0.25rem;
  background-color: #FFFFFF;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  color: #222222;
  height: 2rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

<div
  aria-hidden="true"
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      class="c2"
      focusable="false"
      height="12px"
      viewBox="0 0 32 32"
      width="12px"
    >
      <polygon
        points="3,32 29,16 3,0"
      />
    </svg>
    <span
      class="c3"
    >
      Video
    </span>
  </div>
</div>
`;

exports[`MediaIndicator should render video correctly without duration details 1`] = `
.c3 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
}

.c2 {
  vertical-align: middle;
  margin: 0 0.25rem;
  fill: #222222;
  width: 0.75rem;
  height: 0.75rem;
}

.c0 {
  padding: 0.5rem 0.25rem;
  background-color: #FFFFFF;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  color: #222222;
  height: 2rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

<div
  aria-hidden="true"
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      class="c2"
      focusable="false"
      height="12px"
      viewBox="0 0 32 32"
      width="12px"
    >
      <polygon
        points="3,32 29,16 3,0"
      />
    </svg>
    <span
      class="c3"
    >
      Video
    </span>
  </div>
</div>
`;

exports[`MediaIndicator should render video indicator correctly 1`] = `
.c3 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
}

.c2 {
  vertical-align: middle;
  margin: 0 0.25rem;
  fill: #222222;
  width: 0.75rem;
  height: 0.75rem;
}

.c0 {
  padding: 0.5rem 0.25rem;
  background-color: #FFFFFF;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  color: #222222;
  height: 2rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

.c4 {
  vertical-align: middle;
  margin: 0 0.25rem;
}

<div
  aria-hidden="true"
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      class="c2"
      focusable="false"
      height="12px"
      viewBox="0 0 32 32"
      width="12px"
    >
      <polygon
        points="3,32 29,16 3,0"
      />
    </svg>
    <span
      class="c3"
    >
      Video
    </span>
    <time
      class="c4"
      datetime="PT2M15S"
    >
      2:15
    </time>
  </div>
</div>
`;
